runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/fn/control/UserDistributionMonitoringInfoToCounterUpdateTransformer.java

```diff
@@ -0,0 +1,143 @@
+      final Map<String, DataflowStepContext> transformIdMapping) {
+    this.transformIdMapping = transformIdMapping;
+    this.specValidator = specMonitoringInfoValidator;
+  }
+
+  static final String BEAM_METRICS_USER_DISTRIBUTION_PREFIX =
+      Enum.USER_DISTRIBUTION_COUNTER
+          .getValueDescriptor()
+          .getOptions()
+          .getExtension(monitoringInfoSpec)
+          .getUrn();
+
+  private Optional<String> validate(MonitoringInfo monitoringInfo) {
+    Optional<String> validatorResult = specValidator.validate(monitoringInfo);
+    if (validatorResult.isPresent()) {
+      return validatorResult;
+    }
+
+    String urn = monitoringInfo.getUrn();
+    if (!urn.startsWith(BEAM_METRICS_USER_DISTRIBUTION_PREFIX)) {
+      throw new RuntimeException(
+          String.format(
+              "Received unexpected counter urn. Expected urn starting with: %s, received: %s",
+              BEAM_METRICS_USER_DISTRIBUTION_PREFIX, urn));
+    }
+
+    final String ptransform = monitoringInfo.getLabelsMap().get("PTRANSFORM");
+    DataflowStepContext stepContext = transformIdMapping.get(ptransform);
+    if (stepContext == null) {
+      return Optional.of(
+          "Encountered user-counter MonitoringInfo with unknown ptransformId: "
+              + monitoringInfo.toString());
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * Transforms user counter MonitoringInfo to relevant CounterUpdate.
+   *
+   * @return Relevant CounterUpdate or null if transformation failed.
+   */
+  @Override
+  public CounterUpdate transform(MonitoringInfo monitoringInfo) {
+    Optional<String> validationResult = validate(monitoringInfo);
+    if (validationResult.isPresent()) {
+      LOG.info(validationResult.get());
+      return null;
+    }
+
+    IntDistributionData value =
+        monitoringInfo.getMetric().getDistributionData().getIntDistributionData();
+    String urn = monitoringInfo.getUrn();
+
+    final String ptransform = monitoringInfo.getLabelsMap().get("PTRANSFORM");
+
+    CounterStructuredNameAndMetadata name = new CounterStructuredNameAndMetadata();
+
+    String nameWithNamespace =
+        urn.substring(BEAM_METRICS_USER_DISTRIBUTION_PREFIX.length()).replace("^:", "");
+
+    final int lastColonIndex = nameWithNamespace.lastIndexOf(':');
+    String counterName = nameWithNamespace.substring(lastColonIndex + 1);
+    String counterNamespace =
+        lastColonIndex == -1 ? "" : nameWithNamespace.substring(0, lastColonIndex);
+
+    DataflowStepContext stepContext = transformIdMapping.get(ptransform);
+    name.setName(
+            new CounterStructuredName()
+                .setOrigin(Origin.USER.toString())
+                .setName(counterName)
+                .setOriginalStepName(stepContext.getNameContext().originalName())
+                .setOriginNamespace(counterNamespace))
+        .setMetadata(new CounterMetadata().setKind("DISTRIBUTION"));
```

Comment:
Just to clarify - Dataflow considers "distribution" to be a _kind_ of "counter"?

Reply:
yes.